File: engines/python/agentce/store.py
"""The evidence graph store (ADR-0001): an embedded, on-disk SQLite database.

The provenance graph is held as triples in relational tables -- object-property edges, datatype
literals, and the transitive ``rdfs:subClassOf*`` closure -- written once during graph building and
queried read-mostly during evaluation. Every query orders explicitly, so results never depend on
physical row order (HR-1 determinism). PSP shapes compile to SQL over these tables (item 1.7).

IRIs are stored as compact CURIEs (``agentce:``, ``prov:``, ``rdf:type``) rather than expanded URLs,
which keeps the tables small and comparisons exact.
"""

from __future__ import annotations

import sqlite3
from collections.abc import Iterable
from pathlib import Path

RDF_TYPE = "rdf:type"

_SCHEMA = """
CREATE TABLE IF NOT EXISTS edges (
    s TEXT NOT NULL, p TEXT NOT NULL, o TEXT NOT NULL,
    PRIMARY KEY (s, p, o)
) WITHOUT ROWID;
CREATE INDEX IF NOT EXISTS edges_po ON edges (p, o);
CREATE INDEX IF NOT EXISTS edges_sp ON edges (s, p);
CREATE TABLE IF NOT EXISTS literals (
    s TEXT NOT NULL, p TEXT NOT NULL, val TEXT NOT NULL, datatype TEXT NOT NULL,
    PRIMARY KEY (s, p, val, datatype)
) WITHOUT ROWID;
CREATE INDEX IF NOT EXISTS literals_sp ON literals (s, p);
CREATE TABLE IF NOT EXISTS class_closure (
    descendant TEXT NOT NULL, ancestor TEXT NOT NULL,
    PRIMARY KEY (descendant, ancestor)
) WITHOUT ROWID;
"""


class GraphStore:
    """A SQLite-backed triple store for the evidence graph."""

    def __init__(self, path: Path | str = ":memory:") -> None:
        self.conn = sqlite3.connect(str(path))
        self.conn.executescript(_SCHEMA)

    def close(self) -> None:
        self.conn.close()

    def __enter__(self) -> GraphStore:
        return self

    def __exit__(self, *exc: object) -> None:
        self.close()

    # --- writes (idempotent; INSERT OR IGNORE keeps triples a set) ---

    def add_edge(self, s: str, p: str, o: str) -> None:
        self.conn.execute(
            "INSERT OR IGNORE INTO edges (s, p, o) VALUES (?, ?, ?)", (s, p, o)
        )

    def add_type(self, s: str, cls: str) -> None:
        self.add_edge(s, RDF_TYPE, cls)

    def add_literal(
        self, s: str, p: str, val: str, datatype: str = "xsd:string"
    ) -> None:
        self.conn.execute(
            "INSERT OR IGNORE INTO literals (s, p, val, datatype) VALUES (?, ?, ?, ?)",
            (s, p, val, datatype),
        )

    def add_subclass_closure(self, pairs: Iterable[tuple[str, str]]) -> None:
        self.conn.executemany(
            "INSERT OR IGNORE INTO class_closure (descendant, ancestor) VALUES (?, ?)",
            list(pairs),
        )

    def commit(self) -> None:
        self.conn.commit()

    # --- reads (always ordered) ---

    def objects(self, s: str, p: str) -> list[str]:
        rows = self.conn.execute(
            "SELECT o FROM edges WHERE s = ? AND p = ? ORDER BY o", (s, p)
        ).fetchall()
        return [row[0] for row in rows]

    def subjects(self, p: str, o: str) -> list[str]:
        rows = self.conn.execute(
            "SELECT s FROM edges WHERE p = ? AND o = ? ORDER BY s", (p, o)
        ).fetchall()
        return [row[0] for row in rows]

    def literal_values(self, s: str, p: str) -> list[str]:
        rows = self.conn.execute(
            "SELECT val FROM literals WHERE s = ? AND p = ? ORDER BY val", (s, p)
        ).fetchall()
        return [row[0] for row in rows]

    def instances_of(self, cls: str) -> list[str]:
        """Every node whose type is ``cls`` or a subclass of it (via the materialised closure)."""
        rows = self.conn.execute(
            """
            SELECT DISTINCT e.s FROM edges e
            JOIN class_closure c ON e.o = c.descendant
            WHERE e.p = ? AND c.ancestor = ?
            ORDER BY e.s
            """,
            (RDF_TYPE, cls),
        ).fetchall()
        return [row[0] for row in rows]

    def is_a(self, node: str, cls: str) -> bool:
        row = self.conn.execute(
            """
            SELECT 1 FROM edges e
            JOIN class_closure c ON e.o = c.descendant
            WHERE e.s = ? AND e.p = ? AND c.ancestor = ?
            LIMIT 1
            """,
            (node, RDF_TYPE, cls),
        ).fetchone()
        return row is not None

    def edge_count(self) -> int:
        return int(self.conn.execute("SELECT COUNT(*) FROM edges").fetchone()[0])

    def literal_count(self) -> int:
        return int(self.conn.execute("SELECT COUNT(*) FROM literals").fetchone()[0])

    def triple_count(self) -> int:
        """Total triples (object edges plus datatype literals)."""
        return self.edge_count() + self.literal_count()

File: engines/typescript/src/store.ts
/**
 * The evidence graph store (ADR-0001), as an in-memory triple set.
 *
 * The Python reference holds the graph in SQLite; this engine holds the same triples in memory and
 * answers the same queries in the same order (every read is sorted by UTF-8 bytes, matching SQLite's
 * BINARY collation) so the two engines build byte-identical evidence. Triples are a set: writes are
 * idempotent. IRIs are compact CURIEs (`agentce:`, `prov:`, `rdf:type`).
 */

import { byteCompare } from "./util";

export const RDF_TYPE = "rdf:type";
const SEP = "";

export class GraphStore {
  /** key `s\x1fp` -> set of objects */
  private readonly spo = new Map<string, Set<string>>();
  /** key `p\x1fo` -> set of subjects */
  private readonly pos = new Map<string, Set<string>>();
  /** key `s\x1fp` -> set of `val\x1fdatatype` */
  private readonly lit = new Map<string, Set<string>>();
  /** `descendant\x1fancestor` pairs */
  private readonly closure = new Set<string>();
  /** ancestor -> set of descendants */
  private readonly descendantsByAncestor = new Map<string, Set<string>>();
  private edges = 0;
  private literals = 0;

  private static add(index: Map<string, Set<string>>, key: string, value: string): boolean {
    let bucket = index.get(key);
    if (bucket === undefined) {
      bucket = new Set<string>();
      index.set(key, bucket);
    }
    if (bucket.has(value)) {
      return false;
    }
    bucket.add(value);
    return true;
  }

  addEdge(s: string, p: string, o: string): void {
    const added = GraphStore.add(this.spo, s + SEP + p, o);
    if (added) {
      GraphStore.add(this.pos, p + SEP + o, s);
      this.edges += 1;
    }
  }

  addType(s: string, cls: string): void {
    this.addEdge(s, RDF_TYPE, cls);
  }

  addLiteral(s: string, p: string, val: string, datatype = "xsd:string"): void {
    const added = GraphStore.add(this.lit, s + SEP + p, val + SEP + datatype);
    if (added) {
      this.literals += 1;
    }
  }

  addSubclassClosure(pairs: Iterable<[string, string]>): void {
    for (const [descendant, ancestor] of pairs) {
      this.closure.add(descendant + SEP + ancestor);
      GraphStore.add(this.descendantsByAncestor, ancestor, descendant);
    }
  }

  objects(s: string, p: string): string[] {
    return [...(this.spo.get(s + SEP + p) ?? [])].sort(byteCompare);
  }

  subjects(p: string, o: string): string[] {
    return [...(this.pos.get(p + SEP + o) ?? [])].sort(byteCompare);
  }

  literalValues(s: string, p: string): string[] {
    const bucket = this.lit.get(s + SEP + p);
    if (bucket === undefined) {
      return [];
    }
    return [...bucket].map((entry) => entry.slice(0, entry.indexOf(SEP))).sort(byteCompare);
  }

  literalPairs(s: string, p: string): Array<[string, string]> {
    const bucket = this.lit.get(s + SEP + p);
    if (bucket === undefined) {
      return [];
    }
    return [...bucket]
      .map((entry): [string, string] => {
        const at = entry.indexOf(SEP);
        return [entry.slice(0, at), entry.slice(at + 1)];
      })
      .sort((a, b) => byteCompare(a[0], b[0]) || byteCompare(a[1], b[1]));
  }

  instancesOf(cls: string): string[] {
    const descendants = this.descendantsByAncestor.get(cls);
    if (descendants === undefined) {
      return [];
    }
    const found = new Set<string>();
    for (const descendant of descendants) {
      for (const subject of this.pos.get(RDF_TYPE + SEP + descendant) ?? []) {
        found.add(subject);
      }
    }
    return [...found].sort(byteCompare);
  }

  isA(node: string, cls: string): boolean {
    for (const type of this.spo.get(node + SEP + RDF_TYPE) ?? []) {
      if (this.closure.has(type + SEP + cls)) {
        return true;
      }
    }
    return false;
  }

  /** Every triple as a sorted line list, for byte-identity checks against the reference engine. */
  dumpTriples(): string[] {
    const lines: string[] = [];
    for (const [key, objects] of this.spo) {
      const [s, p] = key.split(SEP);
      for (const o of objects) {
        lines.push(`E\t${s}\t${p}\t${o}`);
      }
    }
    for (const [key, entries] of this.lit) {
      const [s, p] = key.split(SEP);
      for (const entry of entries) {
        const [val, datatype] = entry.split(SEP);
        lines.push(`L\t${s}\t${p}\t${val}\t${datatype}`);
      }
    }
    for (const pair of this.closure) {
      const [descendant, ancestor] = pair.split(SEP);
      lines.push(`C\t${descendant}\t${ancestor}`);
    }
    return lines.sort(byteCompare);
  }

  edgeCount(): number {
    return this.edges;
  }

  literalCount(): number {
    return this.literals;
  }

  tripleCount(): number {
    return this.edges + this.literals;
  }
}
